### PR TITLE
Recipe app

### PR DESCRIPTION
**Pull request by tumainimaganiko**
In this project, @lordksix  and @tumainimaganiko we did the following:

- We followed the layout of the wireframes provided and personalized the rest of the design including colors, typographies, spacings, etc.
- Login page and registration page:
    - Is built with Devise.
- Food list :
    - It displays a list of food added by the logged-in user as in the wireframe (display also quantity of a given food).
    - It leads to a form that allows users to add new food.
- Recipes list:
    - It displays a list of recipes created by the logged-in user as in the wireframe.
    - It leads to recipe details.
    - If the user is the owner of the recipe, It allows the user to delete it.
- Public recipe list:
    - It displays a list of all public recipes ordered by the newest as in the wireframe.
    - Should lead to recipe details.
- Recipe details:
    - It displays a toggle button that allows for a recipe to be made public or private.
    - If the recipe is public or the user is the owner of the recipe, it displays the recipe details as in the wireframe.
    - If the user is the owner of the recipe, it leads them to the form that allows the user to add new food.
- General shopping list view 
    - It shows the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
    - It counts the total number of food items and the total price of the missing food.
- We made sure there were no N+1 queries happening.
- Created a navigation menu allowing users to open all the pages I created.
- Write unit and integration tests

### CODE REVIEW

File: app/models/food.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class Food < ApplicationRecord
+  belongs_to :user
+
```
Comment (t-yanick) [inline app/models/food.rb:3]: - [x] Great job working on this capstone project. As per the Entity Relationship diagram, you are missing the association between Food and Recipe Food and Recipe Food. I think the following associations exist:
```
  has_many :recipe_foods
  has_many :recipes, through: :recipe_foods
```

File: db/migrate/20230904180116_create_users.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class CreateUsers < ActiveRecord::Migration[7.0]
+  def change
+    create_table :users do |t|
+      t.string :email, null: false, default: ""
```
Comment (t-yanick) [inline db/migrate/20230904180116_create_users.rb]: - [x] As per Entity Relationship diagram, the column in the User table is `name`. But I see that you are using `email`. This needs to be fixed

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <title>RecipeApp</title>
+    <meta name="viewport" content="width=device-width,initial-scale=1">
+    <%= csrf_meta_tags %>
+    <%= csp_meta_tag %>
+    <%= stylesheet_link_tag "tailwind", "inter-font", "data-turbo-track": "reload" %>
```
Comment (t-yanick) [inline app/views/layouts/application.html.erb:8]: - [x] An error is raised here when I run your application locally as seen in the screenshot below. rails db:migrate does not run your migrations. Kindly check that also from your end.
image

image

File: spec/models/food_spec.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
+require 'rails_helper'
+
```
Comment (t-yanick) [inline spec/models/food_spec.rb:2]: - [x] When I run `rspec`, there are 9 failures. Kindly check the screenshot below and fix the failing tests.
image

**Review by t-yanick — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED

Hi @tumainimaganiko and @lordksix , 

### Good job so far!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Use of correct Gitflow
:heavy_check_mark: login and registration build with Devise
:heavy_check_mark:  RSpec and Capybara used for testing
:heavy_check_mark: All necessary routes added
:heavy_check_mark: Correct columns and datatypes setup
:heavy_check_mark: Most associations between models setup correctly
:heavy_check_mark: Use of strong parameters

There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️

- [x] Kindly fix the error resulting from the asset pipeline, so that your app can run. All technical aspects were reviewed, but have not reviewed the display of your app.

- [x] One of the most effective ways to address N+1 queries is to use eager loading. Eager loading allows you to load associated records in a single query, reducing the number of queries executed. You can use the `includes` or preload methods to specify the associations to be eagerly loaded. But you seem not to have used any effective way to address N+1. For example:

**N+1 query example**
```
@users = User.all
@users.each do |user|
  puts user.posts.count
end
```

**Eager loading solution**
```
@users = User.includes(:posts).all
@users.each do |user|
  puts user.posts.count
end
```
- [x] If you only need specific attributes from associated records, you can use the pluck or select methods to fetch only those attributes. This can help avoid loading entire objects and reduce the number of queries. For example:
**N+1 query example**
```
@users = User.all
@users.each do |user|
  puts user.posts.pluck(:title)
end
```

**Pluck solution**
```
@users = User.all
post_titles = Post.where(user_id: @users).pluck(:user_id, :title)
post_titles.each do |user_id, title|
  puts title
end
```

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/layouts/application.html.erb (hunk 3 above)
Comment (lordksix) [reply app/views/layouts/application.html.erb:8]: Thank you for the feedback. To use a tailwind base rails app you need to use `./bin/dev`. If you want to use `rails s`, you need before the first time you run the app you must use this command: `bundle exec rake assets:precompile`

Screenshot from 2023-09-09 00-17-10

Comment (lordksix) [reply app/views/layouts/application.html.erb:8]: Added instructions on the readme for future users.

File: spec/models/food_spec.rb (hunk 4 above)
Comment (lordksix) [reply spec/models/food_spec.rb:2]: The origin of the error is the same as before. Run first  `bundle exec rake assets:precompile`. Instructions were updated in the readme.md file

Screenshot from 2023-09-09 00-15-51

File: db/migrate/20230904180116_create_users.rb (hunk 2 above)
Comment (lordksix) [reply db/migrate/20230904180116_create_users.rb]: Thank you for the feed back. It is fixed.

File: app/views/layouts/application.html.erb (hunk 3 above)
Comment (lordksix) [reply app/views/layouts/application.html.erb:8]: FIXED

File: spec/models/food_spec.rb (hunk 4 above)
Comment (lordksix) [reply spec/models/food_spec.rb:2]: FIXED

File: app/models/food.rb (hunk 1 above)
Comment (lordksix) [reply app/models/food.rb:3]: Thank you for the feedback. We have fixed it.

**Review by elisha2kyakpo1 — APPROVED** [review]
Hi 👋 @tumainimaganiko, @lordksix, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

👏 You did a great job fixing all the issues pointed out by the previous reviewer, the app works as expected.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**